clash-prelude/tests/Clash/Tests/MaybeConvert.hs

```diff
@@ -0,0 +1,323 @@
+>          m = "(m + 1)" if b == "Index" else "m"
+>          print(f"""case_maybeConvert{a}{b} :: Assertion
+> case_maybeConvert{a}{b} =
+>   forM_ [0 .. {ia_max}] $ \\n ->
+>     forM_ [0 .. {ib_max}] $ \\m ->
+>       withSomeSNat n $ \(SNat :: SNat n) ->
+>         withSomeSNat m $ \(SNat :: SNat m) ->
+>           forM_ [minBound .. maxBound] $ \(i :: {a} {n}) -> do
+>             assertBool (show (n, m, i)) (convertMaybeLaw1 (Proxy @({b} {m})) i)
+>             assertBool (show (n, m, i)) (convertMaybeLaw2 (Proxy @({b} {m})) i)
+>             assertBool (show (n, m, i)) (convertMaybeLaw3 (Proxy @({b} {m})) i)
+>             assertBool (show (n, m, i)) (convertMaybeLaw4 (Proxy @({b} {m})) i)
+>  """)
+-}
+module Clash.Tests.MaybeConvert where
+
+import Clash.Class.Convert (MaybeConvert (maybeConvert))
+import Control.Monad (forM_)
+import Data.Data (Proxy (..))
+import Data.Either (isLeft)
+import Data.Maybe (fromMaybe, isJust)
+import GHC.TypeNats (someNatVal)
+import Test.Tasty (TestTree, defaultMain)
+import Test.Tasty.HUnit (Assertion, assertBool, testCase)
+import Test.Tasty.TH (testGroupGenerator)
+
+#if MIN_VERSION_base(4,18,0)
+import Clash.Prelude hiding (someNatVal, withSomeSNat)
+#else
+import Clash.Prelude hiding (someNatVal)
+#endif
+
+#if !MIN_VERSION_base(4,16,0)
+import Numeric.Natural (Natural)
+#endif
+
+main :: IO ()
+main = defaultMain tests
+
+tests :: TestTree
+tests = $(testGroupGenerator)
+
+withSomeSNat :: Natural -> (forall (n :: Nat). SNat n -> r) -> r
+withSomeSNat n f = case someNatVal n of
+  SomeNat (_ :: Proxy n) -> f (SNat @n)
+
+convertMaybeLaw1 ::
+  forall a b.
+  (Eq a, MaybeConvert a b, MaybeConvert b a) =>
+  Proxy b ->
+  a ->
+  Bool
+convertMaybeLaw1 Proxy x =
+  x == fromMaybe x (maybeConvert @_ @b x >>= maybeConvert)
+
+convertMaybeLaw2 ::
+  forall a b.
+  (MaybeConvert a b, MaybeConvert b a, Integral b, Integral a) =>
+  Proxy b ->
+  a ->
+  Bool
+convertMaybeLaw2 Proxy x =
+  toInteger x == fromMaybe (toInteger x) (toInteger <$> maybeConvert @_ @b x)
+
+convertMaybeLaw3 ::
+  forall a b.
+  (MaybeConvert a b, MaybeConvert b a, Integral b, Integral a) =>
+  Proxy b ->
+  a ->
+  Bool
+convertMaybeLaw3 Proxy x =
+  isJust (maybeConvert @_ @b x) `implies` isJust (maybeConvert @_ @a =<< maybeConvert @_ @b x)
+ where
+  implies :: Bool -> Bool -> Bool
+  implies True False = False
+  implies _ _ = True
+
+convertMaybeLaw4 ::
+  forall a b.
+  (MaybeConvert a b, MaybeConvert b a, Integral b, Integral a, Bounded b, Bounded a) =>
+  Proxy b ->
+  a ->
+  Bool
+convertMaybeLaw4 Proxy x =
+  isJust (maybeConvert @_ @b x) == (i x >= i (minBound @b) && i x <= i (maxBound @b))
+ where
+  i :: (Integral c) => c -> Integer
+  i = toInteger
+
+-- | Checks whether an @XException@ in, means an @XException@ out
+convertXException :: forall a b. (MaybeConvert a b) => Proxy a -> Proxy b -> Bool
+convertXException _ _ = isLeft $ isX $ maybeConvert @a @b (errorX "" :: a)
```

Comment:
If you make it so that the original `XException` is always preserved, you could make this test stricter by testing that the message of the original `XException` is also the message of the ouput `XException` (same for the `Convert` tests).

Reply:
Test amended!